### ISSUE TITLE
Suggestion to make change in corpus data so that it can improve accuracy

### ISSUE THREAD

**Issue by vishnu-ks** [gunthercox/ChatterBot#417]
I just made a chatbot with training data that I made by my own. It is trained with a set of conversation about animals which we see commonly. The only problem that what I found is most of the time when we ask the bot which it don't know it will reply an answer which has closest match.
Suppose I am asking the bot 'what is a cat' and it know the response for' what is a rat' only then it reply the answer for what is a rat. So the answer which is given by the bot become entirely wrong. I tried using LowConfidence adapter to tackle with this issue and then I realized I will not be able to get the auto correction feature that currently AIML don't have if I filter response with Low confidence adapter. By auto correction what I meant is sometimes we write 'wat' instead of what, dat instead of 'that' etc. all this can be managed correctly by using closest match adapter. 
So as a solution for all this problem what I was thinking was like if we have a json file of following format as a training set, can we go through each node for searching for keyword and if a keyword is found then can we stick on that specific list of conversation to  find out a closest match response?
{
	"cow": [
		"What is a cow?",
		"Cow is a domestic animal"
	],

	"cat": [
		"What is a cat?",
		"Cat is a pet"
	]
}

Suppose I am asking a question **"what is a cat"** after removing the stopwords or by tokenizing each word we can search and find a node which has a best match and can find a response for the question from the list of corresponding node. Can we do like that?

**Comment by gunthercox** [comment]
So your recommendation is to compare the text of each statement with the stop words removed when searching for the closest match? Please correct me if I'm wrong, I just want to make sure I'm understanding correctly.

**Comment by vishnu-ks** [comment]
I will explain with an example, If a user asks **What is a cat** first we will tokenize it or we remove the stopwords and then tokenize it. Ater tokenizing we will get 4 tokens **What, is, a, cat** and then we will compare all this 4 tokens one by one to all the keys that we have in our json data because we don't know what is the keyword yet. Then after finding the key from the json data we will search for a closest match for the question from that particular node. Then we will get an accurate answer.

**Comment by alekaizer** [comment]
@vishnu-ks I am facing the same kind of issues on the responses, being fighting with chatterbot for 2 weeks now to get the right combination to knock it out :D 

and I slightly have the same idea, using tokenized data as tags then store them in the extra_data field along side the statements.

I will implement it and let @gunthercox see it and maybe he'll give us another hint, he might have an illumination from it and come up with a better idea :)

**Comment by gunthercox** [comment]
Hi @vishnu-ks, @alekaizer, I'm going to close this issue because I'm still uncertain that it actually solves anything. I welcome any pull requests or code samples as those may provide better grounds for demonstrating and collaborating on a solution.
